images/encryption/blockcipher/blockcipher_siv.go

```diff
@@ -0,0 +1,217 @@
+		// for anything less we assume it's EOF
+		numbytes := 0
+		for numbytes < r.bc.toread {
+			var n int
+			n, r.bc.err = r.bc.reader.ReadAt(r.bc.inbuffer[numbytes:r.bc.toread-numbytes], r.bc.inoffset+int64(numbytes))
+			numbytes += n
+			if r.bc.err != nil {
+				if r.bc.err == io.EOF {
+					r.bc.eof = true
+					r.bc.err = nil
+					break
+				} else {
+					return 0, r.bc.err
+				}
+			}
+			if n == 0 {
+				break
+			}
+		}
+		if numbytes < r.bc.toread {
+			r.bc.eof = true
+		}
+
+		r.bc.inoffset += int64(numbytes)
+
+		// transform the data
+		if r.bc.encryptor != nil {
+			r.bc.outbuffer = r.bc.encryptor.Seal(nil, r.bc.inbuffer[:numbytes], []byte(""), r.bc.eof)
+		} else {
+			r.bc.outbuffer, r.bc.err = r.bc.decryptor.Open(nil, r.bc.inbuffer[:numbytes], []byte(""), r.bc.eof)
+			if r.bc.err != nil {
+				return 0, r.bc.err
+			}
+		}
+		// let reader start from beginning of buffer
+		r.bc.outoffset = 0
+		r.bc.outsize += int64(len(r.bc.outbuffer))
+	}
+}
+
+// Size returns the Size of the encrypted data
+func (r *aessivcryptor) Size() int64 {
+	return r.bc.outsize
+}
+
+func (r *aessivcryptor) Digest() digest.Digest {
+	return r.digester.Digest()
+}
+
+func (r *aessivcryptor) Hash() hash.Hash {
+	return r.digester.Hash()
+}
+
+// init initializes an instance
+func (bc *AESSIVLayerBlockCipher) init(encrypt bool, reader io.ReaderAt, opt LayerBlockCipherOptions) (LayerBlockCipherOptions, error) {
+	var (
+		err error
+		se  miscreant.StreamEncryptor
+	)
+
+	bc.reader = reader
+
+	key := opt.SymmetricKey
+	if len(key) != bc.bits/8 {
+		return LayerBlockCipherOptions{}, fmt.Errorf("invalid key length of %d bytes; need %d bytes", len(key), bc.bits/8)
+	}
+
+	nonce := opt.CipherOptions["nonce"]
+	if len(nonce) == 0 {
+		nonce = make([]byte, se.NonceSize())
+		if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+			return LayerBlockCipherOptions{}, errors.Wrap(err, "Unable to generate random nonce")
+		}
+	}
+
+	bc.inbuffer = make([]byte, 1024*1024)
+	bc.toread = len(bc.inbuffer)
+	bc.inoffset = 0
+	bc.outbuffer = nil
+	bc.outoffset = 0
+	bc.eof = false
+	bc.err = nil
+	bc.outsize = 0
+
+	if encrypt {
+		bc.encryptor, err = miscreant.NewStreamEncryptor("AES-SIV", key, nonce)
+		if err != nil {
+			return LayerBlockCipherOptions{}, errors.Wrap(err, "Unable to create AES-SIV stream encryptor")
+		}
+		bc.toread -= bc.encryptor.Overhead()
+		bc.decryptor = nil
+	} else {
+		bc.decryptor, err = miscreant.NewStreamDecryptor("AES-SIV", key, nonce)
+		if err != nil {
+			return LayerBlockCipherOptions{}, errors.Wrap(err, "Unable to create AES-SIV stream decryptor")
+		}
+		bc.encryptor = nil
+	}
+
+	lbco := LayerBlockCipherOptions{
+		SymmetricKey: key,
+		CipherOptions: map[string][]byte{
+			"nonce": nonce,
+		},
+	}
+
+	return lbco, nil
+}
+
+// GenerateKey creates a synmmetric key
+func (bc *AESSIVLayerBlockCipher) GenerateKey() []byte {
+	return miscreant.GenerateKey(bc.bits / 8)
```

Comment:
same as comment above, put key length into struct rather than having divide by 8 in multiple places.

Reply:
Done.